File: src/commands/mcp.rs
use std::fs;
use std::path::PathBuf;

use serde::{Deserialize, Serialize};
use serde_json::{Map, Value, json};

use crate::api::ApiError;
use crate::commands::auth::get_api_key;
use crate::config::Credentials;
use crate::output::{OutputFormat, print_json, print_message};

#[derive(Debug, Serialize, Deserialize, Default)]
#[serde(rename_all = "camelCase")]
struct ClaudeConfig {
    #[serde(default)]
    mcp_servers: Map<String, Value>,
    #[serde(flatten)]
    other: Map<String, Value>,
}

fn get_claude_config_path() -> Result<PathBuf, ApiError> {
    #[cfg(target_os = "macos")]
    {
        let home = dirs::home_dir()
            .ok_or_else(|| ApiError::ConfigError("Could not determine home directory".into()))?;
        Ok(home.join("Library/Application Support/Claude/claude_desktop_config.json"))
    }

    #[cfg(target_os = "windows")]
    {
        let appdata = dirs::config_dir()
            .ok_or_else(|| ApiError::ConfigError("Could not determine AppData directory".into()))?;
        Ok(appdata.join("Claude/claude_desktop_config.json"))
    }

    #[cfg(target_os = "linux")]
    {
        let config = dirs::config_dir()
            .ok_or_else(|| ApiError::ConfigError("Could not determine config directory".into()))?;
        Ok(config.join("Claude/claude_desktop_config.json"))
    }

    #[cfg(not(any(target_os = "macos", target_os = "windows", target_os = "linux")))]
    {
        Err(ApiError::ConfigError("Unsupported platform".into()))
    }
}

pub fn setup(force: bool, format: OutputFormat) -> Result<(), ApiError> {
    let creds = Credentials::load()?;
    let token = get_api_key(&creds).ok_or_else(|| {
        ApiError::Unauthorized(
            "Not logged in. Run 'vector auth login' to authenticate.".to_string(),
        )
    })?;

    let config_path = get_claude_config_path()?;

    // Load existing config or create new one
    let mut config: ClaudeConfig = if config_path.exists() {
        let content = fs::read_to_string(&config_path)
            .map_err(|e| ApiError::ConfigError(format!("Failed to read Claude config: {}", e)))?;
        serde_json::from_str(&content)
            .map_err(|e| ApiError::ConfigError(format!("Failed to parse Claude config: {}", e)))?
    } else {
        ClaudeConfig::default()
    };

    // Check if vector is already configured
    if config.mcp_servers.contains_key("vector") && !force {
        return Err(ApiError::ConfigError(
            "Vector MCP server already configured. Use --force to overwrite.".to_string(),
        ));
    }

    // Create the Vector MCP server configuration
    let vector_config = json!({
        "command": "npx",
        "args": [
            "-y",
            "mcp-remote",
            "https://api.builtfast.com/mcp/vector",
            "--header",
            format!("Authorization: Bearer {}", token)
        ]
    });

    let was_updated = config.mcp_servers.contains_key("vector");

    // Add or update the vector server
    config
        .mcp_servers
        .insert("vector".to_string(), vector_config);

    // Ensure parent directory exists
    if let Some(parent) = config_path.parent()
        && !parent.exists()
    {
        fs::create_dir_all(parent).map_err(|e| {
            ApiError::ConfigError(format!("Failed to create Claude config directory: {}", e))
        })?;
    }

    // Write the config
    let content = serde_json::to_string_pretty(&config)
        .map_err(|e| ApiError::ConfigError(format!("Failed to serialize config: {}", e)))?;
    fs::write(&config_path, content)
        .map_err(|e| ApiError::ConfigError(format!("Failed to write Claude config: {}", e)))?;

    let action = if was_updated { "updated" } else { "added" };

    if format == OutputFormat::Json {
        print_json(&json!({
            "success": true,
            "action": action,
            "config_path": config_path.to_string_lossy(),
            "message": format!("Vector MCP server {} in Claude Desktop config", action)
        }));
    } else {
        print_message(&format!(
            "Vector MCP server {} in Claude Desktop config.",
            action
        ));
        print_message(&format!("Config written to: {}", config_path.display()));
        print_message("\nRestart Claude Desktop to apply changes.");
    }

    Ok(())
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_claude_config_empty() {
        let config: ClaudeConfig = serde_json::from_str("{}").unwrap();
        assert!(config.mcp_servers.is_empty());
        assert!(config.other.is_empty());
    }

    #[test]
    fn test_claude_config_preserves_other_mcp_servers() {
        let json = r#"{
            "mcpServers": {
                "other-server": {
                    "command": "node",
                    "args": ["server.js"]
                }
            }
        }"#;

        let mut config: ClaudeConfig = serde_json::from_str(json).unwrap();
        assert!(config.mcp_servers.contains_key("other-server"));

        // Add vector
        config
            .mcp_servers
            .insert("vector".to_string(), json!({"command": "npx"}));

        // Serialize and deserialize
        let serialized = serde_json::to_string(&config).unwrap();
        let restored: ClaudeConfig = serde_json::from_str(&serialized).unwrap();

        assert!(restored.mcp_servers.contains_key("other-server"));
        assert!(restored.mcp_servers.contains_key("vector"));
    }

    #[test]
    fn test_claude_config_preserves_other_fields() {
        let json = r#"{
            "mcpServers": {},
            "theme": "dark",
            "someOtherSetting": true
        }"#;

        let config: ClaudeConfig = serde_json::from_str(json).unwrap();
        assert!(config.other.contains_key("theme"));
        assert!(config.other.contains_key("someOtherSetting"));

        // Serialize back
        let serialized = serde_json::to_string(&config).unwrap();
        assert!(serialized.contains("theme"));
        assert!(serialized.contains("someOtherSetting"));
    }

    #[test]
    fn test_vector_config_structure() {
        let token = "test-token-123";
        let vector_config = json!({
            "command": "npx",
            "args": [
                "-y",
                "mcp-remote",
                "https://api.builtfast.com/mcp/vector",
                "--header",
                format!("Authorization: Bearer {}", token)
            ]
        });

        assert_eq!(vector_config["command"], "npx");
        let args = vector_config["args"].as_array().unwrap();
        assert_eq!(args[0], "-y");
        assert_eq!(args[1], "mcp-remote");
        assert_eq!(args[2], "https://api.builtfast.com/mcp/vector");
        assert_eq!(args[3], "--header");
        assert_eq!(args[4], "Authorization: Bearer test-token-123");
    }

    #[test]
    fn test_claude_config_roundtrip() {
        let original = r#"{
            "mcpServers": {
                "existing": {"command": "test"}
            },
            "customField": "value"
        }"#;

        let mut config: ClaudeConfig = serde_json::from_str(original).unwrap();
        config
            .mcp_servers
            .insert("vector".to_string(), json!({"command": "npx"}));

        let serialized = serde_json::to_string_pretty(&config).unwrap();
        let restored: ClaudeConfig = serde_json::from_str(&serialized).unwrap();

        assert_eq!(restored.mcp_servers.len(), 2);
        assert!(restored.mcp_servers.contains_key("existing"));
        assert!(restored.mcp_servers.contains_key("vector"));
        assert!(restored.other.contains_key("customField"));
    }

    #[test]
    fn test_get_claude_config_path() {
        let path = get_claude_config_path().unwrap();
        assert!(path.ends_with("claude_desktop_config.json"));
        assert!(path.to_string_lossy().contains("Claude"));
    }
}

File: internal/commands/mcp.go
package commands

import (
	"encoding/json"
	"errors"
	"fmt"
	"os"
	"path/filepath"
	"runtime"

	"github.com/spf13/cobra"

	"github.com/built-fast/vector-cli/internal/api"
)

const mcpServerURL = "https://api.builtfast.com/mcp/vector"

// claudeDesktopConfigPathFn is the function used to resolve the desktop config path.
// Override in tests for deterministic paths.
var claudeDesktopConfigPathFn = claudeDesktopConfigPath

// NewMcpCmd creates the mcp command group.
func NewMcpCmd() *cobra.Command {
	cmd := &cobra.Command{
		Use:   "mcp",
		Short: "MCP server configuration",
		Long:  "Configure the Vector MCP server for use with Claude Desktop or Claude Code.",
	}

	cmd.AddCommand(newMcpSetupCmd())

	return cmd
}

func newMcpSetupCmd() *cobra.Command {
	var target string
	var global bool
	var force bool

	cmd := &cobra.Command{
		Use:   "setup",
		Short: "Configure Vector MCP server",
		Long:  "Configure the Vector MCP server in Claude Desktop or Claude Code for AI-assisted site management.",
		RunE: func(cmd *cobra.Command, args []string) error {
			app, err := requireApp(cmd)
			if err != nil {
				return err
			}

			token := app.Client.Token

			// Validate --global only with --target code
			if global && target != "code" {
				return fmt.Errorf("--global flag only applies when --target is \"code\"")
			}

			// Determine config path
			configPath, err := mcpConfigPath(target, global)
			if err != nil {
				return err
			}

			// Build the MCP server entry
			serverEntry := buildMcpServerEntry(target, token)

			// Read existing config or start fresh
			configData, err := readJSONFile(configPath)
			if err != nil {
				return err
			}

			// Get or create mcpServers
			mcpServers, _ := configData["mcpServers"].(map[string]any)
			if mcpServers == nil {
				mcpServers = map[string]any{}
			}

			// Check if already configured
			action := "added"
			if _, exists := mcpServers["vector"]; exists {
				if !force {
					return &api.APIError{
					Message:  "Vector MCP server already configured. Use --force to overwrite.",
					ExitCode: 1,
				}
				}
				action = "updated"
			}

			// Set the vector entry
			mcpServers["vector"] = serverEntry
			configData["mcpServers"] = mcpServers

			// Create parent directories
			dir := filepath.Dir(configPath)
			if err := os.MkdirAll(dir, 0o755); err != nil {
				return fmt.Errorf("failed to create directory %s: %w", dir, err)
			}

			// Write config
			data, err := json.MarshalIndent(configData, "", "  ")
			if err != nil {
				return fmt.Errorf("failed to marshal config: %w", err)
			}
			data = append(data, '\n')

			if err := os.WriteFile(configPath, data, 0o644); err != nil {
				return fmt.Errorf("failed to write config: %w", err)
			}

			// Success messages
			w := cmd.OutOrStdout()
			targetLabel := "Claude Desktop"
			if target == "code" {
				targetLabel = "Claude Code"
			}
			_, _ = fmt.Fprintf(w, "Vector MCP server %s in %s config.\n", action, targetLabel)
			_, _ = fmt.Fprintf(w, "Config written to: %s\n", configPath)

			// Restart message: omitted for Code project-level config
			if target == "desktop" || global {
				_, _ = fmt.Fprintf(w, "Restart %s to apply changes.\n", targetLabel)
			}

			return nil
		},
	}

	cmd.Flags().StringVar(&target, "target", "desktop", "Target application: \"desktop\" or \"code\"")
	cmd.Flags().BoolVar(&global, "global", false, "Write to global config (~/.claude.json) instead of project-level .mcp.json (only for --target code)")
	cmd.Flags().BoolVar(&force, "force", false, "Overwrite existing Vector MCP configuration")

	return cmd
}

// mcpConfigPath returns the config file path for the given target.
func mcpConfigPath(target string, global bool) (string, error) {
	switch target {
	case "desktop":
		return claudeDesktopConfigPathFn()
	case "code":
		if global {
			home, err := os.UserHomeDir()
			if err != nil {
				return "", fmt.Errorf("unable to determine home directory: %w", err)
			}
			return filepath.Join(home, ".claude.json"), nil
		}
		return ".mcp.json", nil
	default:
		return "", fmt.Errorf("invalid target %q: must be \"desktop\" or \"code\"", target)
	}
}

// claudeDesktopConfigPath returns the platform-specific Claude Desktop config path.
func claudeDesktopConfigPath() (string, error) {
	switch runtime.GOOS {
	case "darwin":
		home, err := os.UserHomeDir()
		if err != nil {
			return "", fmt.Errorf("unable to determine home directory: %w", err)
		}
		return filepath.Join(home, "Library", "Application Support", "Claude", "claude_desktop_config.json"), nil
	case "linux":
		home, err := os.UserHomeDir()
		if err != nil {
			return "", fmt.Errorf("unable to determine home directory: %w", err)
		}
		return filepath.Join(home, ".config", "Claude", "claude_desktop_config.json"), nil
	case "windows":
		appData := os.Getenv("APPDATA")
		if appData == "" {
			return "", fmt.Errorf("%%APPDATA%% is not set")
		}
		return filepath.Join(appData, "Claude", "claude_desktop_config.json"), nil
	default:
		return "", fmt.Errorf("unsupported platform: %s", runtime.GOOS)
	}
}

// buildMcpServerEntry returns the MCP server config for the given target.
func buildMcpServerEntry(target, token string) map[string]any {
	entry := map[string]any{
		"command": "npx",
		"args": []string{
			"-y",
			"mcp-remote",
			mcpServerURL,
			"--header",
			fmt.Sprintf("Authorization: Bearer %s", token),
		},
	}
	if target == "code" {
		entry["type"] = "stdio"
	}
	return entry
}

// readJSONFile reads a JSON file into a map. Returns an empty map if the file doesn't exist.
func readJSONFile(path string) (map[string]any, error) {
	data, err := os.ReadFile(path)
	if err != nil {
		if errors.Is(err, os.ErrNotExist) {
			return map[string]any{}, nil
		}
		return nil, fmt.Errorf("failed to read %s: %w", path, err)
	}

	var result map[string]any
	if err := json.Unmarshal(data, &result); err != nil {
		return nil, fmt.Errorf("failed to parse %s: %w", path, err)
	}

	return result, nil
}
